core/cas-server-core-services/src/main/java/org/apereo/cas/services/DefaultRegisteredServiceAccessStrategy.java

```diff
@@ -230,91 +230,63 @@ public boolean doPrincipalAttributesAllowServiceAccess(final String principal, f
         }
         
-        if (this.requireAllAttributes && copy.size() < this.requiredAttributes.size()) {
+        if (this.requireAllAttributes && difference.size() < this.requiredAttributes.size()) {
             LOGGER.debug("Not all required attributes are available to the principal");
             return false;
         }
         
-        return copy.stream().anyMatch(key -> {
+        return difference.stream().anyMatch(key -> {
             final Set<String> requiredValues = this.requiredAttributes.get(key);
-            final Set<String> availableValues;
+            final Set<String> availableValues = objectToSetOfStrings(principalAttributes.get(key));
 
-            final Object objVal = principalAttributes.get(key);
-            if (objVal instanceof Collection) {
-                availableValues = Sets.newHashSet(((Collection) objVal).iterator());
-            } else {
-                availableValues = Sets.newHashSet(objVal.toString());
-            }
-
-            final Set<?> differenceInValues;
             final Pattern pattern = RegexUtils.concatenate(requiredValues, this.caseInsensitive);
             if (pattern != RegexUtils.MATCH_NOTHING_PATTERN) {
-                final Predicate<String> patternPredicate = pattern.asPredicate();
-                differenceInValues = availableValues.stream().filter(patternPredicate).collect(Collectors.toSet());
+                return availableValues.stream().anyMatch(pattern.asPredicate());
             } else {
-                differenceInValues = Sets.intersection(availableValues, requiredValues);
-            }
-
-            if (!differenceInValues.isEmpty()) {
-                LOGGER.info("Principal is authorized to access the service");
-                return true;
+                return availableValues.stream().anyMatch(requiredValues::contains);
             }
-            return false;
         });
     }
     
     private boolean doRejectedAttributesRefusePrincipalAccess(final Map<String, Object> principalAttributes) {
         LOGGER.debug("These rejected attributes [{}] are examined against [{}] before service can proceed.",
                 this.rejectedAttributes, principalAttributes);
-        final Sets.SetView<String> rejectedDifference = Sets.intersection(this.rejectedAttributes.keySet(), principalAttributes.keySet());
-        final Set<String> rejectedCopy = rejectedDifference.immutableCopy();
+        final Set<String> rejectedDifference = this.rejectedAttributes.keySet().stream()
+                .filter(a -> principalAttributes.keySet().contains(a))
+                .collect(Collectors.toSet());
 
         if (this.rejectedAttributes.isEmpty()) {
             LOGGER.debug("No rejected attributes are defined");
             return false;
         }
         
-        if (this.requireAllAttributes && rejectedCopy.size() < this.rejectedAttributes.size()) {
+        if (this.requireAllAttributes && rejectedDifference.size() < this.rejectedAttributes.size()) {
```

Comment:
It seems like the two blocks for rejectedDifference and difference collections are almost identical. Can we refactor and reuse?